diff -ruN ata.orig/ata-all.c ata/ata-all.c
--- ata.orig/ata-all.c	2007-11-09 10:58:07.000000000 +0100
+++ ata/ata-all.c	2008-02-07 12:30:21.000000000 +0100
@@ -78,6 +78,7 @@
 uma_zone_t ata_request_zone;
 uma_zone_t ata_composite_zone;
 int ata_wc = 1;
+int ata_to = 5;
 
 /* local vars */
 static int ata_dma = 1;
@@ -94,6 +95,9 @@
 TUNABLE_INT("hw.ata.wc", &ata_wc);
 SYSCTL_INT(_hw_ata, OID_AUTO, wc, CTLFLAG_RDTUN, &ata_wc, 0,
 	   "ATA disk write caching");
+TUNABLE_INT("hw.ata.to", &ata_to);
+SYSCTL_INT(_hw_ata, OID_AUTO, to, CTLFLAG_RW, &ata_to, 0,
+	   "ATA disk timeout vis-a-vis power-saving");
 
 /*
  * newbus device interface related functions
diff -ruN ata.orig/ata-all.h ata/ata-all.h
--- ata.orig/ata-all.h	2007-11-22 08:50:36.000000000 +0100
+++ ata/ata-all.h	2008-02-07 12:31:35.000000000 +0100
@@ -519,6 +519,7 @@
 extern struct intr_config_hook *ata_delayed_attach;
 extern devclass_t ata_devclass;
 extern int ata_wc;
+extern int ata_to;
  
 /* public prototypes */
 /* ata-all.c: */
diff -ruN ata.orig/ata-disk.c ata/ata-disk.c
--- ata.orig/ata-disk.c	2007-12-09 20:27:19.000000000 +0100
+++ ata/ata-disk.c	2008-02-07 12:32:06.000000000 +0100
@@ -230,7 +230,7 @@
     request->dev = dev;
     request->bio = bp;
     request->callback = ad_done;
-    request->timeout = 5;
+    request->timeout = ata_to;
     request->retries = 2;
     request->data = bp->bio_data;
     request->bytecount = bp->bio_bcount;
--- ata/ata-chipset.c	2008-10-21 18:40:29.000000000 +0000
+++ ata/ata-chipset.c	2009-04-27 17:51:39.000000000 +0000
@@ -5410,6 +5410,7 @@
      { ATA_VIA8237A,  0x00, VIA133, 0x00,    ATA_UDMA6, "8237A" },
      { ATA_VIA8237S,  0x00, VIA133, 0x00,    ATA_UDMA6, "8237S" },
      { ATA_VIA8251,   0x00, VIA133, 0x00,    ATA_UDMA6, "8251" },
+     { ATA_VIACX700M2,   0x00, VIA133, 0x00,    ATA_UDMA6, "CX700M2" },
      { 0, 0, 0, 0, 0, 0 }};
     static struct ata_chip_id new_ids[] =
     {{ ATA_VIA6410,   0x00, 0,      0x00,    ATA_UDMA6, "6410" },
@@ -5418,6 +5419,7 @@
      { ATA_VIA8237A,  0x00, 7,      0x00,    ATA_SA150, "8237A" },
      { ATA_VIA8237S,  0x00, 7,      0x00,    ATA_SA150, "8237S" },
      { ATA_VIA8251,   0x00, 0,      VIAAHCI, ATA_SA300, "8251" },
+     { ATA_VIACX700M2,   0x00, 0,      0x00, ATA_SA300, "CX700M2" },
      { 0, 0, 0, 0, 0, 0 }};
 
     if (pci_get_devid(dev) == ATA_VIA82C571) {
--- ata/ata-pci.h	2008-10-02 02:57:24.000000000 +0000
+++ ata/ata-pci.h	2009-04-27 17:50:22.000000000 +0000
@@ -379,6 +379,7 @@
 #define ATA_VIA6410             0x31641106
 #define ATA_VIA6420             0x31491106
 #define ATA_VIA6421             0x32491106
+#define ATA_VIACX700M2		0x53241106
 
 /* chipset setup related defines */
 #define AHCI            1
