*** pmap.h.orig	2009-12-07 13:56:56.000000000 +0100
--- pmap.h	2009-12-07 13:57:16.000000000 +0100
***************
*** 115,121 ****
  
  /* Initial number of kernel page tables. */
  #ifndef NKPT
! #define	NKPT		32
  #endif
  
  #define NKPML4E		1		/* number of kernel PML4 slots */
--- 115,121 ----
  
  /* Initial number of kernel page tables. */
  #ifndef NKPT
! #define	NKPT		60
  #endif
  
  #define NKPML4E		1		/* number of kernel PML4 slots */
