diff -ruN fusefs-ntfs.orig/Makefile fusefs-ntfs/Makefile
--- fusefs-ntfs.orig/Makefile	Tue Apr 24 14:21:41 2007
+++ fusefs-ntfs/Makefile	Tue Apr 24 11:21:56 2007
@@ -5,7 +5,7 @@
 #
 
 PORTNAME=	ntfs
-PORTVERSION=	1.0
+PORTVERSION=	1.417
 CATEGORIES=	sysutils
 MASTER_SITES=	http://www.ntfs-3g.org/
 MASTER_SITE_SUBDIR=	fuse
diff -ruN fusefs-ntfs.orig/distinfo fusefs-ntfs/distinfo
--- fusefs-ntfs.orig/distinfo	Tue Apr 24 14:21:41 2007
+++ fusefs-ntfs/distinfo	Tue Apr 24 14:26:38 2007
@@ -1,3 +1,3 @@
-MD5 (ntfs-3g-1.0.tgz) = 873a8de662849d129fc7c475ad3f5447
-SHA256 (ntfs-3g-1.0.tgz) = b5bbf81230bb21573d69fc471721b0b0c404eaa5040ef28e7dcec38320a1b2e4
-SIZE (ntfs-3g-1.0.tgz) = 562452
+MD5 (ntfs-3g-1.417.tgz) = 9e451f1084461795a96170adb6c4376b
+SHA256 (ntfs-3g-1.417.tgz) = 6eff152c670cb9cab1c3d423aac1ce7825f3bc5f53be5ff1857c0079bd27ecc0
+SIZE (ntfs-3g-1.417.tgz) = 562094
diff -ruN fusefs-ntfs.orig/files/patch-Makefile.in fusefs-ntfs/files/patch-Makefile.in
--- fusefs-ntfs.orig/files/patch-Makefile.in	Tue Apr 24 14:21:28 2007
+++ fusefs-ntfs/files/patch-Makefile.in	Tue Apr 24 14:51:10 2007
@@ -1,5 +1,5 @@
---- Makefile.in.orig	Wed Dec  6 01:08:22 2006
-+++ Makefile.in	Wed Dec  6 01:08:30 2006
+--- Makefile.in.orig	Mon Apr 16 01:50:20 2007
++++ Makefile.in	Tue Apr 24 11:42:41 2007
 @@ -18,7 +18,7 @@
  VPATH = @srcdir@
  pkgdatadir = $(datadir)/@PACKAGE@
diff -ruN fusefs-ntfs.orig/files/patch-configure fusefs-ntfs/files/patch-configure
--- fusefs-ntfs.orig/files/patch-configure	Tue Apr 24 14:21:28 2007
+++ fusefs-ntfs/files/patch-configure	Tue Apr 24 14:49:12 2007
@@ -1,6 +1,6 @@
---- configure.orig	Tue Jul 25 23:41:16 2006
-+++ configure	Tue Jul 25 23:40:37 2006
-@@ -19956,7 +19956,7 @@
+--- configure.orig	Mon Apr 16 00:50:07 2007
++++ configure	Tue Apr 24 11:38:43 2007
+@@ -19828,7 +19828,7 @@
  
  
  case "$target_os" in
diff -ruN fusefs-ntfs.orig/files/patch-libntfs-3g-device_io.c fusefs-ntfs/files/patch-libntfs-3g-device_io.c
--- fusefs-ntfs.orig/files/patch-libntfs-3g-device_io.c	Tue Apr 24 14:21:28 2007
+++ fusefs-ntfs/files/patch-libntfs-3g-device_io.c	Tue Apr 24 14:58:16 2007
@@ -1,5 +1,5 @@
---- libntfs-3g/device_io.c.orig	Wed Dec  6 04:03:06 2006
-+++ libntfs-3g/device_io.c	Tue Dec  5 23:04:04 2006
+--- libntfs-3g/device_io.c.orig	Tue Apr 24 11:50:00 2007
++++ libntfs-3g/device_io.c	Tue Apr 24 11:49:47 2007
 @@ -26,7 +26,11 @@
  #ifndef __CYGWIN32__
  
diff -ruN fusefs-ntfs.orig/files/patch-src-Makefile.in fusefs-ntfs/files/patch-src-Makefile.in
--- fusefs-ntfs.orig/files/patch-src-Makefile.in	Tue Apr 24 14:21:29 2007
+++ fusefs-ntfs/files/patch-src-Makefile.in	Thu Jan  1 01:00:00 1970
@@ -1,16 +0,0 @@
---- src/Makefile.in.orig	Tue Jan  2 02:21:12 2007
-+++ src/Makefile.in	Thu Jan 11 03:38:55 2007
-@@ -521,11 +521,11 @@
- 
- install-data-am: install-man
- 	@$(NORMAL_INSTALL)
--	$(MAKE) $(AM_MAKEFLAGS) install-data-hook
-+#	$(MAKE) $(AM_MAKEFLAGS) install-data-hook
- 
- install-exec-am: install-binPROGRAMS
- 	@$(NORMAL_INSTALL)
--	$(MAKE) $(AM_MAKEFLAGS) install-exec-hook
-+#	$(MAKE) $(AM_MAKEFLAGS) install-exec-hook
- 
- install-info: install-info-am
- 
