--- /sys/modules/Makefile.orig	2008-08-15 10:40:14.000000000 +0200
+++ /sys/modules/Makefile	2008-08-15 08:57:41.000000000 +0200
@@ -177,6 +177,7 @@
 	${_ncv} \
 	${_ndis} \
 	netgraph \
+	${_nfe} \
 	nfsclient \
 	nfsserver \
 	nge \
@@ -428,6 +429,7 @@
 _ixgb=  	ixgb
 _mly=		mly
 _mxge=		mxge
+_nfe=		nfe
 _nve=		nve
 .if !defined(NO_CRYPT) || defined(ALL_MODULES)
 .if exists(${.CURDIR}/../crypto/via)
@@ -486,6 +488,7 @@
 _mly=		mly
 _mxge=		mxge
 _ndis=		ndis
+_nfe=		nfe
 _nve=		nve
 _ppc=		ppc
 _rr232x=	rr232x
