--- /sys/conf/files.i386.orig	2008-08-15 11:05:00.000000000 +0200
+++ /sys/conf/files.i386	2008-08-15 09:36:27.000000000 +0200
@@ -234,6 +234,7 @@
 dev/mxge/mxge_rss_eth_z8e.c	optional mxge pci
 dev/mxge/mxge_rss_ethp_z8e.c	optional mxge pci
 net/zlib.c			optional mxge
+dev/nfe/if_nfe.c		optional nfe pci
 dev/nve/if_nve.c		optional nve pci
 dev/ppc/ppc.c			optional ppc
 dev/ppc/ppc_puc.c		optional ppc puc pci
--- /sys/conf/files.amd64.orig	2008-08-15 11:04:59.000000000 +0200
+++ /sys/conf/files.amd64	2008-08-15 09:37:02.000000000 +0200
@@ -200,6 +200,7 @@
 dev/mxge/mxge_rss_eth_z8e.c	optional	mxge pci
 dev/mxge/mxge_rss_ethp_z8e.c	optional	mxge pci
 net/zlib.c			optional	mxge
+dev/nfe/if_nfe.c		optional	nfe pci
 dev/nve/if_nve.c		optional	nve pci
 dev/rr232x/os_bsd.c		optional	rr232x
 dev/rr232x/osm_bsd.c		optional	rr232x
