--- nve/if_nve.c.orig	Fri Jan 25 12:30:21 2008
+++ nve/if_nve.c	Fri Jan 25 12:32:49 2008
@@ -237,6 +237,14 @@
 	"NVIDIA nForce MCP12 Networking Adapter"},
 	{NVIDIA_VENDORID, NFORCE_MCPNET13_DEVICEID,
 	"NVIDIA nForce MCP13 Networking Adapter"},
+	{NVIDIA_VENDORID, NFORCE_MCPNET16_DEVICEID,
+	"NVIDIA nForce MCP16 Networking Adapter"},
+	{NVIDIA_VENDORID, NFORCE_MCPNET17_DEVICEID,
+	"NVIDIA nForce MCP17 Networking Adapter"},
+	{NVIDIA_VENDORID, NFORCE_MCPNET18_DEVICEID,
+	"NVIDIA nForce MCP18 Networking Adapter"},
+	{NVIDIA_VENDORID, NFORCE_MCPNET19_DEVICEID,
+	"NVIDIA nForce MCP19 Networking Adapter"},
 	{0, 0, NULL}
 };
 
@@ -490,8 +498,18 @@
 	
 	/* MAC is loaded backwards into h/w reg */
 	sc->hwapi->pfnGetNodeAddress(sc->hwapi->pADCX, sc->original_mac_addr);
-	for (i = 0; i < 6; i++) {
-		eaddr[i] = sc->original_mac_addr[5 - i];
+	if (pci_get_device(dev) == NFORCE_MCPNET19_DEVICEID)
+	{
+		/* Already in the right order */
+		for (i = 0; i < 6; i++) {
+			eaddr[i] = sc->original_mac_addr[i];
+		}
+	}
+	else
+	{
+		for (i = 0; i < 6; i++) {
+			eaddr[i] = sc->original_mac_addr[5 - i];
+		}
 	}
 	sc->hwapi->pfnSetNodeAddress(sc->hwapi->pADCX, eaddr);
 
--- nve/if_nvereg.h.orig	Fri Jan 25 12:28:05 2008
+++ nve/if_nvereg.h	Fri Jan 25 12:28:56 2008
@@ -48,6 +48,12 @@
 #define	NFORCE_MCPNET11_DEVICEID 0x0038 
 #define	NFORCE_MCPNET12_DEVICEID 0x0268
 #define	NFORCE_MCPNET13_DEVICEID 0x0269
+#define	NFORCE_MCPNET14_DEVICEID 0x0372
+#define	NFORCE_MCPNET15_DEVICEID 0x0373
+#define	NFORCE_MCPNET16_DEVICEID 0x03e5
+#define	NFORCE_MCPNET17_DEVICEID 0x03e6
+#define	NFORCE_MCPNET18_DEVICEID 0x03ee
+#define	NFORCE_MCPNET19_DEVICEID 0x03ef
 
 #define	NV_RID		0x10
 
