--- g_raid5.c.orig	2008-07-29 22:58:03.000000000 +0900
+++ g_raid5.c	2011-10-18 01:21:59.000000000 +0900
@@ -46,6 +46,7 @@
 #include <sys/mutex.h>
 #include <sys/bio.h>
 #include <sys/sysctl.h>
+#include <sys/sbuf.h>
 #include <sys/malloc.h>
 #include <sys/proc.h>
 #include <sys/kthread.h>
@@ -2012,7 +2013,7 @@
 	wakeup(&sc->term);
 
 	curthread->td_pflags &= ~TDP_GEOM;
-	kthread_exit(0);
+	kproc_exit(0);
 }
 
 static __inline void
@@ -2441,7 +2442,7 @@
 	wakeup(sc);
 
 	curthread->td_pflags &= ~TDP_GEOM;
-	kthread_exit(0);
+	kproc_exit(0);
 }
 
 /*
@@ -2632,11 +2633,11 @@
 	sc->sc_type = type;
 
 	sc->term = 0;
-	if (kthread_create(g_raid5_worker, sc, &sc->worker, 0, 0,
+	if (kproc_create(g_raid5_worker, sc, &sc->worker, 0, 0,
 							 "g_raid5/main %s", md->md_name) != 0) {
 		sc->workerD = NULL;
 		sc->worker = NULL;
-	} else if (kthread_create(g_raid5_workerD, sc, &sc->workerD, 0, 0,
+	} else if (kproc_create(g_raid5_workerD, sc, &sc->workerD, 0, 0,
 							 "g_raid5/done %s", md->md_name) != 0) {
 		sc->workerD = NULL;
 		sc->term = 1;
